### PR TITLE
feat: enable rate limiting for routes

### LINKED ISSUE #567: Path rewrites should support rate limiting
As continuation of Path Rewrites (#490) feature, rate limiting to routes should be applied.

### Rate Limiting

To ensure reliability and availability of the routes targets (upstreams), it could be valuable to set up rate limits
to the route configurations. The idea is to add `roles.<role_name>.limits.<route_name>` block to the configuration to
define the limit numbers and specify `userRoles` access control section.
This will allow keep up the stability of routes in the face of unexpected load.

For example,
```json
{
  "routes": {
    "vector_store_query": {
      "paths": ["/v1/vector_stores(/[^/]+)*$"],
      "methods": ["GET"],
      "rewritePath": true,
      "userRoles": ["vector_store_user", "vector_store_admin"],
      "upstream": [
        {
          "endpoint": "[https://vector_store.svc.local](https://urldefense.com/v3/__https://vector_store.svc.local__;!!KLCbKzk!hdF9qUPvDZ8EjKZ0gbcvnU7jfUhkET1AabtdFQD99_HZTkfKHrx93o0Ygazrzo-7MeFBBPFrftQODshe_VHvNTnI58qW2LF_$)"
        }
      ]
    },
    "vector_store_mutation": {
      "paths": ["/v1/vector_stores(/[^/]+)*$"],
      "methods": ["POST", "PUT", "DELETE"],
      "rewritePath": true,
      "userRoles": ["vector_store_admin"],
      "upstream": [
        {
          "endpoint": "[https://vector_store.svc.local](https://urldefense.com/v3/__https://vector_store.svc.local__;!!KLCbKzk!hdF9qUPvDZ8EjKZ0gbcvnU7jfUhkET1AabtdFQD99_HZTkfKHrx93o0Ygazrzo-7MeFBBPFrftQODshe_VHvNTnI58qW2LF_$)"
        }
      ]
    }
  },
  "roles": {
    "vector_store_user": {
      "limits": {
        "vector_store_query": {
          "requestHour": "100",
          "requestDay": "1000"
        }
      }
    },
    "vector_store_admin": {
      "limits": {
        "vector_store_query": {},
        "vector_store_mutation": {}
      }
    }
  }
}
```

So, the identities having `vector_store_user` are able to call the service 100 req/min and 1000 req/day,
while `vector_store_admin` have unlimited access.

### PR DESCRIPTION
**Pull request by alekseyvdovenko**
Closes https://github.com/epam/ai-dial-core/issues/567

### CODE REVIEW

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.epam.aidial.core.config;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import lombok.Data;
+
+import java.util.Collection;
+import java.util.Set;
+
+@Data
+public abstract class AccessControlled {
```
Comment (astsiapanay) [inline config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: May be `RoleBasedEntity`? or just `BaseEntity`

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    private String name;
+
+    @JsonAlias({"userRoles", "user_roles"})
+    private Set<String> userRoles;
+
+    public boolean hasAccess(Collection<String> roles) {
```
Comment (astsiapanay) [inline config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: The method confuses a bit. 

Perhaps we need to have better naming for variables. Actually we want to match expected to actual user roles. It's not clear what roles are expected/required and what - actual.

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        if (userRoles == null) {
+            return true;
+        }
+
+        if (userRoles.isEmpty() || roles == null || roles.isEmpty()) {
```
Comment (astsiapanay) [inline config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: Could you please keep the same code as it was? It looks more compact.

```
if (expectedUserRoles == null) {
             return true;
         }

         return !expectedUserRoles.isEmpty()
                 && actualUserRoles.stream().anyMatch(expectedUserRoles::contains);
```

File: server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java [hunk 4]
```diff
@@ -108,7 +110,15 @@ private void handleRequestBody(Buffer requestBody) {
         context.setRequestBody(requestBody);
 
         if (context.getResponseBody() == null) {
-            sendRequest();
+            proxy.getRateLimiter().limit(context, context.getRoute())
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java:113]: You missed a scenario when future result of rateLimitResponse is failed.

File: server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java [hunk 5]
```diff
@@ -254,20 +274,9 @@ private Route selectRoute() {
     @SneakyThrows
     private String getEndpointUri(Upstream upstream) {
         URIBuilder uriBuilder = new URIBuilder(upstream.getEndpoint());
-        if (context.isRewritePath()) {
+        if (context.getRoute() != null && context.getRoute().isRewritePath()) {
```
Comment (astsiapanay) [inline server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java]: `context.getRoute()` might return null?

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java (hunk 2 above)
Comment (astsiapanay) [reply config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: Please add javadoc

**Review by astsiapanay — CHANGES_REQUESTED** [review]

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java (hunk 1 above)
Comment (alekseyvdovenko) [reply config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: renamed to `RoleBasedEntity`

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java (hunk 2 above)
Comment (alekseyvdovenko) [reply config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: added javadoc

File: config/src/main/java/com/epam/aidial/core/config/AccessControlled.java (hunk 3 above)
Comment (alekseyvdovenko) [reply config/src/main/java/com/epam/aidial/core/config/AccessControlled.java]: returned the previous logic

File: server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java (hunk 4 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java:113]: added `onFailure` handler

File: server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java (hunk 5 above)
Comment (alekseyvdovenko) [reply server/src/main/java/com/epam/aidial/core/server/controller/RouteController.java]: yes, must not be null, removed this condition

File: server/src/test/java/com/epam/aidial/core/server/controller/RoleBasedEntityTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package com.epam.aidial.core.server.controller;
```
Comment (astsiapanay) [inline server/src/test/java/com/epam/aidial/core/server/controller/RoleBasedEntityTest.java]: Wrong package. It must be `config`

Comment (alekseyvdovenko) [reply server/src/test/java/com/epam/aidial/core/server/controller/RoleBasedEntityTest.java]: moved to `config`

**Review by astsiapanay — APPROVED** [review]

**Pull request merged**